web3_proxy/src/stats/stat_buffer.rs

```diff
@@ -127,22 +131,83 @@ impl StatBuffer {
                     // trace!("Received stat");
                     // save the stat to a buffer
+
+                    // TODO: tokio spawn this!
                     match stat {
-                        Some(AppStat::RpcQuery(stat)) => {
-                            if self.influxdb_client.is_some() {
-                                // TODO: round the timestamp at all?
+                        Some(AppStat::RpcQuery(request_metadata)) => {
+                            // we convert on this side of the channel so that we don't slow down the request
+                            let stat = RpcQueryStats::try_from_metadata(request_metadata)?;
+
+                            // update the latest balance
+                            // do this BEFORE emitting any stats
+                            let paid_credits_used: bool;
+                            let latest_balance: Balance;
+                            if let Some(db_conn) = self.db_conn.as_ref() {
+                                let user_id = stat.authorization.checks.user_id;
+
+                                // update the user's balance
+                                if user_id == 0 {
+                                    paid_credits_used = false;
+                                    latest_balance = Balance::default();
+                                } else {
+                                    // update the user's cached balance
+                                    let mut user_balance = stat.authorization.checks.latest_balance.write().await;
+
+                                    paid_credits_used = user_balance.active_premium();
+
+                                    user_balance.total_frontend_requests += 1;
+
+                                    if !stat.backend_rpcs_used.is_empty() {
+                                        user_balance.total_cache_misses += 1;
+                                    }
+
+                                    if paid_credits_used {
+                                        // TODO: this lets them get a negative remaining balance. we should clear if close to 0
+                                        user_balance.total_spent_paid_credits += stat.compute_unit_cost;
+
+                                        // check if they still have premium
+                                        if user_balance.active_premium() {
+                                            // TODO: referall credits here? i think in the save_db section still makes sense for those
+                                        } else {
+                                            // was premium, but isn't anymore due to paying for this query. clear the cache
+                                            // TODO: stop at <$0.000001 instead of negative?
+                                            self.user_balance_cache.invalidate(&user_balance.user_id).await;
+
+                                            let rpc_keys = rpc_key::Entity::find()
```

Comment:
could this possibly not catch all the rpc keys? probably should be fine

Reply:
how would it miss one?